File: TimberCruisingApp/InventoryEngine/VolumeEquations/Formfactor.swift
// Central-European metric stem-volume — generic form-factor (Formzahl) volume.
// This is the approximate m³ path for Germany (see
// docs/volume-standards-research.md §C.2).
//
//     V_m3 = g · h · f
//       g = π · (d/100)² / 4     basal area at breast height (m²), d in cm
//       h = total height (m)
//       f = form factor (Formzahl) = actual stem volume / reference cylinder
//
// The whole-tree form factor is "commonly in the order of magnitude of
// 0.45–0.55" (AWF-Wiki, Univ. Göttingen); we default to f = 0.50. This is
// deliberately labelled an APPROXIMATE standard — production-grade German
// volume would call the BDAT taper library (no short closed form exists), but
// the form factor gives a defensible first-order estimate with a real,
// citable formula.

import Foundation

public struct Formfactor: VolumeEquation {

    /// Form factor (Formzahl). Documented range 0.45–0.55; default 0.50.
    public let f: Float
    public let merchFraction: Float

    public init(formFactor: Float = 0.50, merchFraction: Float = 0.85) {
        self.f = formFactor
        self.merchFraction = merchFraction
    }

    /// Factory path — reads an optional `f` override (else 0.50).
    public init(coefficients: [String: Float]) {
        self.f = CoefficientLookup.optional(coefficients, "f", default: 0.50)
        self.merchFraction = CoefficientLookup.optional(coefficients, "merchFraction", default: 0.85)
    }

    public func totalVolumeM3(dbhCm d: Float, heightM h: Float) -> Float {
        guard d > 0, h > 0 else { return 0 }
        let basalAreaM2 = Float.pi * pow(d / 100.0, 2) / 4.0
        return basalAreaM2 * h * f
    }

    public func merchantableVolumeM3(dbhCm: Float, heightM: Float,
                                     topDibCm: Float, stumpHeightCm: Float) -> Float {
        totalVolumeM3(dbhCm: dbhCm, heightM: heightM) * merchFraction
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/inventory/Formfactor.kt
// Metric stem-volume engine — generic form-factor volume for Central-European
// (German) species. An APPROXIMATION, deliberately labelled as such: Germany's
// production standard is the spline-based BDAT taper program, which has no
// short closed form (see docs/volume-standards-research.md §C). The form
// factor gives a usable first-order m³ estimate when a full taper model is
// not bundled.
//
//   V_m³ = g · h · f
//     g = π · (d/100)² / 4   (basal area at breast height, m², d in cm)
//     h = total height, m
//     f = form factor (Formzahl) = actual stem volume / reference-cylinder
//         volume. Whole-tree f is "commonly 0.45–0.55"; default 0.50.
//
// The form factor is carried on the persisted VolumeEquation record's
// dictionary under "f" (overridable per species when better values exist);
// absent, it defaults to 0.50.

package com.hcjeong.forestix.inventory

import kotlin.math.PI

class Formfactor(coefficients: Map<String, Float>) : VolumeEquation {

    /// Whole-tree form factor (Formzahl). Default 0.50 (0.45–0.55 range).
    private val f: Float = CoefficientLookup.optional(coefficients, "f", default = 0.50f)

    private val merchFraction: Float =
        CoefficientLookup.optional(coefficients, "merchFraction", default = 0.85f)

    override fun totalVolumeM3(dbhCm: Float, heightM: Float): Float {
        if (dbhCm <= 0f || heightM <= 0f) return 0f
        val dM = dbhCm / 100f                       // cm → m
        val g = (PI.toFloat() * dM * dM) / 4f       // basal area, m²
        return g * heightM * f
    }

    override fun merchantableVolumeM3(dbhCm: Float, heightM: Float,
                                      topDibCm: Float, stumpHeightCm: Float): Float =
        totalVolumeM3(dbhCm = dbhCm, heightM = heightM) * merchFraction
}
